xdsl/transforms/canonicalize/canonicalize_pass.py

```diff
@@ -0,0 +1,52 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+from xdsl.dialects import builtin
+from xdsl.ir import MLContext, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+)
+from xdsl.traits import HasCanonicalisationPatternsTrait
+from xdsl.transforms.dead_code_elimination import dce
+
+
+@dataclass
+class CanonicalizationPattern(GreedyRewritePatternApplier):
+    """
+    A variant on the GreedyRewritePatternApplier that applies at most
+    n times per op to prevent infinite loops.
+    """
+
+    op_match_count: dict[Operation, int] = field(
+        default_factory=lambda: defaultdict(int)
+    )
+    repeat_apply_limit: int = field(default=5)
```

Comment:
I still think this should be unbounded by default. That should be a bug to have an infinite loop in the canonicalization patterns.